File: app/src/main/java/com/hetica/AutismoCordoba/temporizadorUnico.java
package com.hetica.AutismoCordoba;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.os.Handler;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * The type Temporizador unico.
 */
public class temporizadorUnico extends AppCompatActivity {

    private long mStartTime;
    private CountDownTimer mCountDownTimer;
    private TextView mTextViewCountDown;
    private TextView textAsig;
    private Button Pausa;
    /**
     * The Db.
     */
    AdminSQLiteOpenHelperStats db;
    /**
     * The Bundle.
     */
    Bundle bundle;
    /**
     * The Time string.
     */
    String timeString;
    /**
     * The Asig.
     */
    String asig;
    private Ringtone r;

    int finFlag;

    /**
     * The Boton fin.
     */
    Button botonFin;
    private Button Fin;

    private long mTimeLeftInMillis;
    private boolean mTimerRunning;

    private Button Main;

    Intent siguiente;
    private long then;
    private int longClickDuration = 3000;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_temporizador_unico);

        bundle = getIntent().getExtras();
        db = new AdminSQLiteOpenHelperStats(this, "Stats.db", null, 1);

        mTextViewCountDown=(TextView)findViewById(R.id.textViewCountDown2);
        textAsig=(TextView)findViewById(R.id.textView56);

        botonFin=(Button)findViewById(R.id.button48);
        botonFin.setVisibility(View.INVISIBLE);
        finFlag=1;
        timeString = bundle.getString("time");
        asig=bundle.getString("asig");
        textAsig.setText(asig);
        Log.e("Date Selected", timeString);
        Pausa = (Button)findViewById(R.id.buttonPausa);
        Fin = (Button)findViewById(R.id.buttonFin2);
        readParams();
        mStartTime=Long.parseLong(timeString)*60000;
        mTimeLeftInMillis=mStartTime;

        Main=(Button)findViewById(R.id.button15);

        Pausa.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mTimerRunning) {
                    pauseTimer();
                } else {
                    startTimer();
                }
            }
        });

        Main.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                if (event.getAction() == MotionEvent.ACTION_DOWN) {
                    then = (long) System.currentTimeMillis();
                } else if (event.getAction() == MotionEvent.ACTION_UP) {
                    if ((System.currentTimeMillis() - then) > longClickDuration) {
                        siguiente = new Intent(getBaseContext(), MainActivity.class);
                        mCountDownTimer.cancel();
                        mTimerRunning = false;
                        startActivity(siguiente);
                        System.out.println("Long Click has happened!");
                        return false;
                    } else {
                        /* Implement short click behavior here or do nothing */
                        System.out.println("Short Click has happened...");
                        return false;
                    }
                }
                return true;
            }
        });
        startTimer();
    }

    @Override
    public void onBackPressed() {

    }

    /**
     * Función que maneja el temporizador
     *
     */
    private void startTimer() {
        mCountDownTimer = new CountDownTimer(mTimeLeftInMillis, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                mTimeLeftInMillis = millisUntilFinished;
                updateCountDownText();
            }

            @Override
            public void onFinish() {
                mTimerRunning = false;
                botonFin.setVisibility(View.VISIBLE);
                Fin.setVisibility(View.INVISIBLE);
                showNotification();

                SimpleDateFormat sdf = new SimpleDateFormat("MMddyyyy");
                db.insertData(asig, sdf.format(new Date()), Integer.valueOf(timeString));
                mTextViewCountDown.setText("00:00");
                //final MediaPlayer alarmSound = MediaPlayer.create(this, R.raw.algo);
                Pausa.setVisibility(View.INVISIBLE);
                //alarmSound.start();
            }
        }.start();

        mTimerRunning = true;
        Pausa.setText("Pausa");
    }

    public void pauseTimer() {
        mCountDownTimer.cancel();
        mTimerRunning = false;

        Pausa.setText("Seguir");
    }

    public void finTimer(View view) {
        timeString=String.valueOf( Integer.valueOf(timeString) - (int) (mTimeLeftInMillis / 1000) / 60);
        mCountDownTimer.cancel();
        mTimerRunning = false;
        finFlag=1;
        SimpleDateFormat sdf = new SimpleDateFormat("MMddyyyy");
        db.insertData(asig, sdf.format(new Date()), Integer.valueOf(timeString));
        Intent siguiente1 = new Intent(this, MainActivity.class);

        startActivity(siguiente1);

    }

    /**
     * Función que escribe el tiempo restante
     *
     */
    private void updateCountDownText() {
        int minutes = (int) (mTimeLeftInMillis / 1000) / 60;
        int seconds = (int) (mTimeLeftInMillis / 1000) % 60;

        String timeLeftFormatted = String.format(Locale.getDefault(), "%02d:%02d", minutes, seconds);

        mTextViewCountDown.setText(timeLeftFormatted);

    }

    /**
     * Función que regresa al menú principal
     *
     * @param view the view
     */
    public void pasarInit(View view)
    {
        Intent siguiente = new Intent(this, MainActivity.class);
        r.stop();
        startActivity(siguiente);
    }
    public void readParams (){
        FileInputStream fis = null;
        String pausa;
        try {
            fis = openFileInput("pausa.txt");
            InputStreamReader isr = new InputStreamReader(fis);
            BufferedReader br = new BufferedReader(isr);
            StringBuilder sb = new StringBuilder();
            String text;

            pausa = br.readLine();



            if (pausa.equalsIgnoreCase("0")){
                Pausa.setVisibility(View.INVISIBLE);

            }




        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (fis != null) {
                try {
                    fis.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }

        String fin;
        try {
            fis = openFileInput("fin.txt");
            InputStreamReader isr = new InputStreamReader(fis);
            BufferedReader br = new BufferedReader(isr);
            StringBuilder sb = new StringBuilder();
            String text;

            fin = br.readLine();



            if (fin.equalsIgnoreCase("0")){
                Fin.setVisibility(View.INVISIBLE);

            }




        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (fis != null) {
                try {
                    fis.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public void showNotification(){

        try {
            Uri notification = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM);
            r = RingtoneManager.getRingtone(getApplicationContext(), notification);
            r.play();
            Handler handler = new Handler();
            handler.postDelayed(new Runnable() {
                public void run() {
                    r.stop();
                }
            }, 60000);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/hetica/AutismoCordoba/temporizadorUnico.kt
package com.hetica.AutismoCordoba


import android.annotation.SuppressLint
import android.content.Intent
import android.media.Ringtone
import android.media.RingtoneManager
import android.os.Bundle
import android.os.CountDownTimer
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.util.Log
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.widget.Button
import android.widget.TextView
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import java.io.BufferedReader
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStreamReader
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * The type Temporizador unico.
 */
class temporizadorUnico : AppCompatActivity() {
    private var mStartTime: Long = 0
    private var mCountDownTimer: CountDownTimer? = null
    private var mTextViewCountDown: TextView? = null
    private var textAsig: TextView? = null
    private var Pausa: Button? = null

    /**
     * The Db.
     */
    var db: AdminSQLiteOpenHelperStats? = null

    /**
     * The Bundle.
     */
    var bundle: Bundle? = null

    /**
     * The Time string.
     */
    var timeString: String? = null

    /**
     * The Asig.
     */
    var asig: String? = null
    private var r: Ringtone? = null
    var finFlag = 0

    /**
     * The Boton fin.
     */
    var botonFin: Button? = null
    private var Fin: Button? = null
    private var mTimeLeftInMillis: Long = 0
    private var mTimerRunning = false
    private var Main: Button? = null
    var siguiente: Intent? = null
    private var then: Long = 0
    var doubleBackToExitPressedOnce = false


    @SuppressLint("MissingInflatedId")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_temporizador_unico)
        bundle = intent.extras
        db = AdminSQLiteOpenHelperStats(this)
        onBackPressedDispatcher.addCallback(this,onBackPressedCallback)
        mTextViewCountDown = findViewById<View>(R.id.textViewCountDown2) as TextView
        textAsig = findViewById<View>(R.id.textView56) as TextView
        botonFin = findViewById<View>(R.id.button48) as Button //el ultimo en aparecer y el que corta la alarma y sale de la actividad
        botonFin!!.visibility = View.INVISIBLE


        finFlag = 1
        timeString = bundle!!.getString("time")
        asig = bundle!!.getString("asig")
        textAsig!!.text = asig
        Log.e("Minutes Selected", timeString!!)
        Pausa = findViewById<View>(R.id.buttonPausa) as Button
        Fin = findViewById<View>(R.id.buttonFin2) as Button //el que es visible de primeras y lleva el marcador al 0, activando la alarma
        readParams()
        mStartTime = timeString!!.toLong() * 60000
        mTimeLeftInMillis = mStartTime
        Main = findViewById<View>(R.id.button15) as Button
        Pausa!!.setOnClickListener {
            if (mTimerRunning) {
                pauseTimer()
            } else {
                startTimer()
            }
        }

        GoToMain()
        startTimer()
    }

    @SuppressLint("ClickableViewAccessibility")
    fun GoToMain() {
        var isLongPressFired = false
        val handler = Handler(Looper.getMainLooper())
        val delayMillis = 3000L // 3 segundos
        val gestureDetector = GestureDetector(this, object : GestureDetector.SimpleOnGestureListener() {
            override fun onLongPress(e: MotionEvent) {
                if (isLongPressFired) {
                    return
                }
                isLongPressFired = true

                // Usamos un Handler para retrasar la apertura de la actividad EditarCalendario
                handler.postDelayed({
                    val intent = Intent(this@temporizadorUnico, MainActivity::class.java)
                    startActivity(intent)
                }, delayMillis)
            }
        })

        Main?.setOnTouchListener { _, event ->
            gestureDetector.onTouchEvent(event)
            if (event.action == MotionEvent.ACTION_UP || event.action == MotionEvent.ACTION_CANCEL) {
                isLongPressFired = false
                // Si se libera el botón antes del tiempo de espera, cancelamos el Handler
                handler.removeCallbacksAndMessages(null)
                mCountDownTimer?.cancel()
            }
            true
        }
    }


    private val onBackPressedCallback: OnBackPressedCallback = object : OnBackPressedCallback(true) {
        override fun handleOnBackPressed() {
            if (doubleBackToExitPressedOnce) {
                val siguiente = Intent(baseContext, MainActivity::class.java)
                startActivity(siguiente)
                mCountDownTimer?.cancel()
                finish()
                return
            }
            doubleBackToExitPressedOnce = true
            FuncionesComunes.showSnackbarWithCustomTextSize(
                this@temporizadorUnico,
                "Presiona de nuevo para salir.",
            )
            Handler(Looper.getMainLooper()).postDelayed({
                doubleBackToExitPressedOnce = false
            }, 2000)
            // Override if needed

        }
    }
    override fun onPause() {
        super.onPause()
        // Detener la alarma si está sonando
        r?.stop()
    }
    /**
     * Función que maneja el temporizador
     *
     */
    private fun startTimer() {
        mCountDownTimer = object : CountDownTimer(mTimeLeftInMillis, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                mTimeLeftInMillis = millisUntilFinished
                updateCountDownText()
            }

            override fun onFinish() {
                mTimerRunning = false
                Main!!.visibility=View.INVISIBLE
                botonFin!!.visibility = View.VISIBLE
                Fin!!.visibility = View.INVISIBLE
                showNotification()
                val sdf = SimpleDateFormat("dd/MM/yyyy")
                db!!.insertData(asig!!, sdf.format(Date()), Integer.valueOf(timeString!!))
                mTextViewCountDown!!.text = "00:00"
                //final MediaPlayer alarmSound = MediaPlayer.create(this, R.raw.algo);
                Pausa!!.visibility = View.INVISIBLE
                //alarmSound.start();
            }
        }.start()
        mTimerRunning = true
        Pausa!!.text = "Pausa"
    }

    fun pauseTimer() {
        mCountDownTimer!!.cancel()
        mTimerRunning = false
        Pausa!!.text = "Seguir"
    }

    fun finTimer(view: View?) {
        // Deshabilita el segundo botón Finalizar y Pausa para evitar múltiples clics
        Fin!!.isEnabled = false
        Pausa!!.isEnabled=false
        timeString?.let {
            val intValue = Integer.valueOf(it) - (mTimeLeftInMillis / 1000).toInt() / 60
            timeString = intValue.toString()
        }
        mCountDownTimer!!.cancel()
        mTimerRunning = false
        Main!!.visibility=View.INVISIBLE
        botonFin!!.visibility = View.VISIBLE
        Fin!!.visibility = View.INVISIBLE
        Pausa!!.visibility=View.INVISIBLE
        showNotification()
        val sdf = SimpleDateFormat("dd/MM/yyyy")
        db!!.insertData(asig!!, sdf.format(Date()), Integer.valueOf(timeString!!))
        mTextViewCountDown!!.text = "00:00"
        mTimeLeftInMillis = 0
        updateCountDownText()

        //basura para que no me de error con por no usar view
        val siguiente2 = Intent(view!!.context, Recompensa::class.java)
        Log.d("hola","$siguiente2")
        //



    }



    /**
     * Función que escribe el tiempo restante
     *
     */
    private fun updateCountDownText() {
        val minutes = (mTimeLeftInMillis / 1000).toInt() / 60
        val seconds = (mTimeLeftInMillis / 1000).toInt() % 60
        val timeLeftFormatted = String.format(Locale.getDefault(), "%02d:%02d", minutes, seconds)
        mTextViewCountDown!!.text = timeLeftFormatted
    }

    /**
     * Función que regresa al menú principal
     *
     * @param view the view
     */
    fun pasarInit(view: View?) {
        val siguiente = Intent(view!!.context, Recompensa::class.java)
        r!!.stop()
        startActivity(siguiente)
    }

    fun readParams() {
        var fis: FileInputStream? = null
        val pausa: String
        try {
            fis = openFileInput("pausa.txt")
            val isr = InputStreamReader(fis)
            val br = BufferedReader(isr)
            StringBuilder()
            pausa = br.readLine()
            if (pausa.equals("0", ignoreCase = true)) {
                Pausa!!.visibility = View.INVISIBLE
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            if (fis != null) {
                try {
                    fis.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
        val fin: String
        try {
            fis = openFileInput("fin.txt")
            val isr = InputStreamReader(fis)
            val br = BufferedReader(isr)
            StringBuilder()
            fin = br.readLine()
            if (fin.equals("0", ignoreCase = true)) {
                Fin!!.visibility = View.INVISIBLE
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            if (fis != null) {
                try {
                    fis.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
    }
    fun showNotification() {
        try {
            val notification = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM)
            r = RingtoneManager.getRingtone(applicationContext, notification)
            r?.play()

            val handlerThread = HandlerThread("StopRingtoneThread")
            handlerThread.start()

            val handler = Handler(handlerThread.looper)
            handler.postDelayed({
                r?.stop()
                handlerThread.quitSafely()
            }, 60000)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

}